File: filefn/server/src/authz/grants.service.ts
import type { Adapter } from '@superfunctions/db';
import type { FileProviderContext } from '@superfunctions/files';
import type { FileRecord } from '../files/service.js';
import type { FilePermissionRecord } from './authorizer.js';
import { randomUUID } from 'crypto';
import * as errors from '../errors.js';

export interface GrantsServiceConfig {
  db: Adapter;
  namespace?: string;
}

export interface CreateGrantInput {
  fileId: string;
  userId?: string;
  role?: string;
  tenantId?: string;
  canRead?: boolean;
  canWrite?: boolean;
  canDelete?: boolean;
  canShare?: boolean;
  expiresAt?: string;
}

export function createGrantsService(config: GrantsServiceConfig) {
  const { db, namespace = 'filefn' } = config;

  async function getFile(fileId: string): Promise<FileRecord | null> {
    return db.findOne<FileRecord>({
      model: 'files',
      where: [{ field: 'fileId', operator: 'eq', value: fileId }],
      namespace,
    });
  }

  async function canManagePermissions(file: FileRecord, ctx: FileProviderContext): Promise<boolean> {
    return file.ownerId === ctx.principalId;
  }

  return {
    async createGrant(input: CreateGrantInput, ctx: FileProviderContext): Promise<FilePermissionRecord> {
      const file = await getFile(input.fileId);
      if (!file) {
        throw errors.notFound('File');
      }

      const canManage = await canManagePermissions(file, ctx);
      if (!canManage) {
        throw errors.forbidden();
      }

      if (!input.userId && !input.role && !input.tenantId) {
        throw new errors.FileFnError('FILEFN_INVALID_GRANT', 'Grant must specify userId, role, or tenantId', 400);
      }

      const permission: FilePermissionRecord = {
        permissionId: randomUUID(),
        fileId: input.fileId,
        userId: input.userId || null,
        role: input.role || null,
        tenantId: input.tenantId || null,
        canRead: input.canRead ?? true,
        canWrite: input.canWrite ?? false,
        canDelete: input.canDelete ?? false,
        canShare: input.canShare ?? false,
        expiresAt: input.expiresAt || null,
        createdAt: new Date().toISOString(),
      };

      await db.create({
        model: 'filePermissions',
        data: permission,
        namespace,
      });

      return permission;
    },

    async listGrants(fileId: string, ctx: FileProviderContext): Promise<FilePermissionRecord[]> {
      const file = await getFile(fileId);
      if (!file) {
        throw errors.notFound('File');
      }

      const canManage = await canManagePermissions(file, ctx);
      if (!canManage) {
        throw errors.forbidden();
      }

      return db.findMany<FilePermissionRecord>({
        model: 'filePermissions',
        where: [{ field: 'fileId', operator: 'eq', value: fileId }],
        namespace,
      });
    },

    async revokeGrant(fileId: string, permissionId: string, ctx: FileProviderContext): Promise<void> {
      const file = await getFile(fileId);
      if (!file) {
        throw errors.notFound('File');
      }

      const canManage = await canManagePermissions(file, ctx);
      if (!canManage) {
        throw errors.forbidden();
      }

      const permission = await db.findOne<FilePermissionRecord>({
        model: 'filePermissions',
        where: [
          { field: 'permissionId', operator: 'eq', value: permissionId },
          { field: 'fileId', operator: 'eq', value: fileId },
        ],
        namespace,
      });

      if (!permission) {
        throw errors.permissionNotFound();
      }

      await db.delete({
        model: 'filePermissions',
        where: [{ field: 'permissionId', operator: 'eq', value: permissionId }],
        namespace,
      });
    },
  };
}

export type GrantsService = ReturnType<typeof createGrantsService>;

File: filefn/python/filefn/server/authz/grants_service.py
import uuid
from datetime import datetime, timezone
from typing import Any, List, Optional

from pydantic import BaseModel

from .. import errors
from ..models import FilePermissionRecord, FileProviderContext, FileRecord


class CreateGrantInput(BaseModel):
    fileId: str
    userId: Optional[str] = None
    role: Optional[str] = None
    tenantId: Optional[str] = None
    canRead: Optional[bool] = None
    canWrite: Optional[bool] = None
    canDelete: Optional[bool] = None
    canShare: Optional[bool] = None
    expiresAt: Optional[str] = None

class GrantsServiceConfig(BaseModel):
    db: Any # Adapter
    namespace: str = 'filefn'

    class Config:
        arbitrary_types_allowed = True

class GrantsService:
    def __init__(self, config: GrantsServiceConfig):
        self.db = config.db
        self.namespace = config.namespace

    async def _get_file(self, file_id: str) -> Optional[FileRecord]:
        data = await self.db.find_one(
            model='files',
            where=[{'field': 'fileId', 'operator': 'eq', 'value': file_id}],
            namespace=self.namespace
        )
        return FileRecord(**data) if data else None

    async def _can_manage_permissions(self, file: FileRecord, ctx: FileProviderContext) -> bool:
        return file.ownerId == ctx.principalId

    async def create_grant(self, input: CreateGrantInput, ctx: FileProviderContext) -> FilePermissionRecord:
        file = await self._get_file(input.fileId)
        if not file:
            raise errors.not_found('File')

        if not await self._can_manage_permissions(file, ctx):
            raise errors.forbidden()

        if not input.userId and not input.role and not input.tenantId:
            raise errors.FileFnError('FILEFN_INVALID_GRANT', 'Grant must specify userId, role, or tenantId', 400)

        permission = FilePermissionRecord(
            permissionId=str(uuid.uuid4()),
            fileId=input.fileId,
            userId=input.userId,
            role=input.role,
            tenantId=input.tenantId,
            canRead=input.canRead if input.canRead is not None else True,
            canWrite=input.canWrite if input.canWrite is not None else False,
            canDelete=input.canDelete if input.canDelete is not None else False,
            canShare=input.canShare if input.canShare is not None else False,
            expiresAt=input.expiresAt,
            createdAt=datetime.now(timezone.utc).isoformat()
        )

        await self.db.create(
            model='filePermissions',
            data=permission.model_dump(),
            namespace=self.namespace
        )

        return permission

    async def list_grants(self, file_id: str, ctx: FileProviderContext) -> List[FilePermissionRecord]:
        file = await self._get_file(file_id)
        if not file:
            raise errors.not_found('File')

        if not await self._can_manage_permissions(file, ctx):
            raise errors.forbidden()

        data = await self.db.find_many(
            model='filePermissions',
            where=[{'field': 'fileId', 'operator': 'eq', 'value': file_id}],
            namespace=self.namespace
        )
        return [FilePermissionRecord(**p) for p in data]

    async def revoke_grant(self, file_id: str, permission_id: str, ctx: FileProviderContext) -> None:
        file = await self._get_file(file_id)
        if not file:
            raise errors.not_found('File')

        if not await self._can_manage_permissions(file, ctx):
            raise errors.forbidden()

        permission = await self.db.find_one(
            model='filePermissions',
            where=[
                {'field': 'permissionId', 'operator': 'eq', 'value': permission_id},
                {'field': 'fileId', 'operator': 'eq', 'value': file_id}
            ],
            namespace=self.namespace
        )

        if not permission:
            raise errors.permission_not_found()

        await self.db.delete(
            model='filePermissions',
            where=[{'field': 'permissionId', 'operator': 'eq', 'value': permission_id}],
            namespace=self.namespace
        )

def create_grants_service(config: GrantsServiceConfig) -> GrantsService:
    return GrantsService(config)
